### PR TITLE
NixOS: add `system.checks`

### PR DESCRIPTION
**Pull request by roberth**
###### Description of changes

Adds an option for configuration checks.
Drops a few unnecessary store paths from the built system configurations.
Manually tested.

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - **tested by hand** because we don't have test infrastructure for asserting about failures in dependencies of derivations.
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by roberth** [comment]
@ofborg test simple

File: nixos/modules/system/activation/top-level.nix [hunk 1]
```diff
@@ -363,14 +378,18 @@ in
           fi
         '';
 
-    system.systemBuilderArgs = lib.optionalAttrs (config.system.forbiddenDependenciesRegex != "") {
+    system.systemBuilderArgs = {
+      passedChecks = concatStringsSep " " config.system.checks;
```
Comment (RaitoBezarius) [inline nixos/modules/system/activation/top-level.nix:382]: Where is `passedChecks` use?

Comment (roberth) [reply nixos/modules/system/activation/top-level.nix:386]: Nowhere. I've added this comment

```
      # Not actually used in the builder. `passedChecks` is just here to create
      # the build dependencies. Checks are similar to build dependencies in the
      # sense that if they fail, the system build fails. However, checks do not
      # produce any output of value, so they are not used by the system builder.
```

I swear that the text alignment is by accident.

Comment (roberth) [reply nixos/modules/system/activation/top-level.nix:389]: And some more
```
      # In fact, using them runs the risk of accidentally adding unneeded paths
      # to the system closure, which defeats the purpose of the `system.checks`
      # option, as opposed to `system.extraDependencies`.
```

Comment (RaitoBezarius) [reply nixos/modules/system/activation/top-level.nix:389]: Awesome as always, Robert!

**Review by RaitoBezarius — APPROVED** [review]
LGTM; **untested yet**

**Comment by roberth** [comment]
> untested yet

Don't worry, I've done that.
Wish we had something for asserting a dependency failure, but alas, manual testing it is.

**Pull request merged**